javascript/ql/src/semmle/javascript/dataflow/internal/CallGraphs.qll

```diff
@@ -0,0 +1,149 @@
+ */
+private import javascript
+
+cached
+module CallGraph {
+  /** Gets the function referenced by `node`, as determined by the type inference. */
+  cached
+  Function getAFunctionValue(AnalyzedNode node) {
+    result = node.getAValue().(AbstractCallable).getFunction()
+  }
+
+  /** Holds if the type inferred for `node` is indefinite due to global flow. */
+  cached
+  predicate isIndefiniteGlobal(AnalyzedNode node) {
+    node.analyze().getAValue().isIndefinite("global")
+  }
+
+  /**
+   * Gets a data flow node that refers to the given function.
+   *
+   * Note that functions are not currently type-tracked, but this exposes the type-tracker `t`
+   * from underlying class tracking if the function came from a class or instance.
+   */
+  pragma[nomagic]
+  private
+  DataFlow::SourceNode getAFunctionReference(DataFlow::FunctionNode function, int imprecision, DataFlow::TypeTracker t) {
+    t.start() and
+    exists(Function fun |
+      fun = function.getFunction() and
+      fun = getAFunctionValue(result)
+    |
+      if isIndefiniteGlobal(result) then
+        fun.getFile() = result.getFile() and imprecision = 0
+        or
+        fun.inExternsFile() and imprecision = 1
+        or
+        imprecision = 2
+      else
+        imprecision = 0
+    )
+    or
+    imprecision = 0 and
+    t.start() and
+    exists(string name |
+      GlobalAccessPath::isAssignedInUniqueFile(name) and
+      GlobalAccessPath::fromRhs(function) = name and
+      GlobalAccessPath::fromReference(result) = name
+    )
+    or
+    imprecision = 0 and
+    exists(DataFlow::ClassNode cls |
+      exists(string name |
+        function = cls.getInstanceMethod(name) and
+        getAnInstanceMemberAccess(cls, name, t.continue()).flowsTo(result)
+        or
+        function = cls.getStaticMethod(name) and
+        cls.getAClassReference(t.continue()).getAPropertyRead(name).flowsTo(result)
+      )
+      or
+      function = cls.getConstructor() and
+      cls.getAClassReference(t.continue()).flowsTo(result)
+    )
+  }
+
+  /**
+   * Gets a data flow node that refers to the given function.
+   */
+  cached
+  DataFlow::SourceNode getAFunctionReference(DataFlow::FunctionNode function, int imprecision) {
+    result = getAFunctionReference(function, imprecision, DataFlow::TypeTracker::end())
+  }
+
+  /**
+   * Gets a data flow node that refers to the result of the given partial function invocation,
+   * with `function` as the underlying function.
+   */
+  pragma[nomagic]
+  private
+  DataFlow::SourceNode getABoundFunctionReference(DataFlow::FunctionNode function, int boundArgs, DataFlow::TypeTracker t) {
+    exists(DataFlow::PartialInvokeNode partial, DataFlow::Node callback |
+      result = partial.getBoundFunction(callback, boundArgs) and
+      getAFunctionReference(function, 0, t.continue()).flowsTo(callback)
+    )
+    or
+    result = getABoundFunctionReferenceAux(function, boundArgs, t)
+  }
+
+  pragma[noopt]
```

Comment:
:cry:

Reply:
Oh, and it's in a recursive predicate, too! Are you absolutely positively sure this is needed?

Reply:
As a concrete suggestion, a quick experiment suggests that you can weaken the `pragma[noopt]` to a `pragma[inline]` like this:

```
  (...)
  or
  exists(DataFlow::StepSummary summary, DataFlow::TypeTracker t2 |
    result = getABoundFunctionReferenceAux(function, boundArgs, t2, summary) and
    t = t2.append(summary)
  )
}

pragma[noinline]
private
DataFlow::SourceNode getABoundFunctionReferenceAux(DataFlow::FunctionNode function, int boundArgs, DataFlow::TypeTracker t, DataFlow::StepSummary summary) {
  exists(DataFlow::SourceNode prev |
    prev = getABoundFunctionReference(function, boundArgs, t) and
    DataFlow::StepSummary::step(prev, result, summary)
  )
}
```

Reply:
Yes I remember it solving a performance problem, but I can run another evaluation with `noinline` if you like.

Reply:
That would be great. I'm not opposed to pragmas on principle, but `pragma[noopt]` is perhaps an overly blunt instrument. In particular, I believe it prevents us from having different orders for different delta sizes, which is generally a good thing. (And yes, it absolutely does solve a performance problem; removing the pragma entirely makes `PostMessageStar` grind to a halt on brackets, for instance.)

Reply:
Results on [nightly](https://git.semmle.com/asger/dist-compare-reports/tree/call-graph-3-alt_1571463031698) look fine. I'm running another eval on default to be sure, but for now I'll merge it in the change.

Reply:
The worker ran out of disk space during the rerun :-\